src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs

```diff
@@ -11,73 +11,154 @@ namespace Microsoft.CodeAnalysis.Differencing
-
-        protected IEnumerable<KeyValuePair<int, int>> GetMatchingPairs(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
+        // VArray struct enables array indexing in range [-d...d].
+        private struct VArray
         {
-            int[,] d = ComputeCostMatrix(oldSequence, oldLength, newSequence, newLength);
-            int i = oldLength;
-            int j = newLength;
+            private readonly int[] _array;
 
-            while (i != 0 && j != 0)
+            public VArray(int d)
             {
-                if (d[i, j] == d[i - 1, j] + DeleteCost)
-                {
-                    i--;
-                }
-                else if (d[i, j] == d[i, j - 1] + InsertCost)
+                _array = new int[2 * d + 1];
+            }
+
+            public void CopyFrom(VArray otherVArray)
+            {
+                int copyDelta = Offset - otherVArray.Offset;
+                if (copyDelta >= 0)
                 {
-                    j--;
+                    Array.Copy(otherVArray._array, 0, _array, copyDelta, otherVArray._array.Length);
                 }
                 else
                 {
-                    i--;
-                    j--;
-                    yield return new KeyValuePair<int, int>(i, j);
+                    Array.Copy(otherVArray._array, -copyDelta, _array, 0, _array.Length);
                 }
             }
-        }
 
-        protected IEnumerable<SequenceEdit> GetEdits(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
-        {
-            int[,] d = ComputeCostMatrix(oldSequence, oldLength, newSequence, newLength);
-            int i = oldLength;
-            int j = newLength;
-
-            while (i != 0 && j != 0)
+            public int this[int index]
             {
-                if (d[i, j] == d[i - 1, j] + DeleteCost)
+                get
                 {
-                    i--;
-                    yield return new SequenceEdit(i, -1);
+                    return _array[index + Offset];
                 }
-                else if (d[i, j] == d[i, j - 1] + InsertCost)
+                set
                 {
-                    j--;
-                    yield return new SequenceEdit(-1, j);
+                    _array[index + Offset] = value;
                 }
-                else
+            }
+
+            private int Offset
+            {
+                get
                 {
-                    i--;
-                    j--;
-                    yield return new SequenceEdit(i, j);
+                    return _array.Length / 2;
                 }
             }
+        }
 
-            while (i > 0)
+        protected abstract bool ItemsEqual(TSequence oldSequence, int oldIndex, TSequence newSequence, int newIndex);
+
+        // TODO: Consolidate return types between GetMatchingPairs and GetEdit to avoid duplicated code
```

Comment:
>TODO: [](start = 11, length = 5)

I created an issue to follow up on this: [https://github.com/dotnet/roslyn/issues/16864](https://github.com/dotnet/roslyn/issues/16864). Please add the link to the comment.